presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java

```diff
@@ -290,6 +290,11 @@ public void addTypes(IdentityHashMap<Expression, Type> types)
         this.types.putAll(types);
     }
 
+    public void replaceType(Expression expression, Type type)
```

Comment:
I don't think you need to replace the type. Just adding the coercion should be enough